kubeflow/kubeflux/all.libsonnet

```diff
@@ -0,0 +1,283 @@
+    $.parts(params, env).role,
+    $.parts(params, env).roleBinding,
+    $.parts(params, env).secret,
+
+  ],
+
+
+  // Parts should be a dictionary containing jsonnet representations of the various
+  // K8s resources used to construct the prototypes listed above.
+  parts(params, env):: {
+    // All ksonnet environments are associated with a namespace and we
+    // generally want to use that namespace for a component.
+    // However, in some cases an application may use multiple namespaces in which
+    // case the namespace for a particular component will be a parameter.
+    local namespace = if std.objectHas(params, "namespace") then params.namespace else env.namespace,
+    local giturl = if std.objectHas(params, "giturl") then params.giturl else env.giturl,
+
+    serviceAccount:: {
+        "apiVersion": "v1",
+        "kind": "ServiceAccount",
+        "metadata": {
+          "labels": {
+            "name": "flux",
+          },
+          "name": "flux"
+        }
+      },
+
+    role:: {
+        "apiVersion": "rbac.authorization.k8s.io/v1beta1",
+        "kind": "ClusterRole",
+        "metadata": {
+          "labels": {
+            "name": "flux",
+          },
+          "name": "flux"
+        },
+        "rules": [
+          {
+            "apiGroups": [
+              "*"
+            ],
+            "resources": [
+              "*"
+            ],
+            "verbs": [
+              "*"
+            ]
+          },
+          {
+            "nonResourceURLs": [
+              "*"
+            ],
+            "verbs": [
+              "*"
+            ]
+          }
+        ]
+      },
+
+      roleBinding:: {
+        "apiVersion": "rbac.authorization.k8s.io/v1beta1",
+        "kind": "ClusterRoleBinding",
+        "metadata": {
+          "labels": {
+            "name": "flux",
+          },
+          "name": "flux"
+        },
+        "roleRef": {
+          "apiGroup": "rbac.authorization.k8s.io",
+          "kind": "ClusterRole",
+          "name": "flux"
+        },
+        "subjects": [
+          {
+            "kind": "ServiceAccount",
+            "name": "flux",
+            "namespace": namespace
+          }
+        ]
+      },
+
+    secret:: {
+      // K8s Deployment,
+      "apiVersion": "v1",
+      "kind": "Secret",
+      "metadata": {
+        "name": "flux-git-deploy",
+      },
+      "type": "Opaque"
+    },
+
+    flux:: {
+      // K8s Deployment,
+      "apiVersion": "apps/v1beta1",
+      "kind": "Deployment",
+      "metadata": {
+        "name": "flux",
+      },
+      "spec": {
+        "replicas": 1,
+        "strategy": {
+          "type": "Recreate"
+        },
+        "template": {
+          "metadata": {
+            "labels": {
+              "name": "flux"
+            }
+          },
+          "spec": {
+            "containers": [
+              {
+                "args": [
+                  "--ssh-keygen-dir=/var/fluxd/keygen",
```

Comment:
How are keys handled? Can the user provide a deploy key via a K8s secret?

Reply:
I don't see why they couldn't create an additional secret for the key. 
I am looking at the documentation here: https://github.com/weaveworks/flux/blob/master/site/standalone/setup.md

There are two options
"fluxctl identity" can generate one after deployment

kubectl create secret generic flux-git-deploy --from-file=identity=/path/to/private_key

Maybe create a parameter for option two?

Reply:
Right; They can create a key but they need some to  specify that the provided secret should be added to the pod. So you probably need a parameter (secret name) and if that is set mount that into the pod.

Reply:
There is an existing "flux-git-deploy" secret created in this process. It looks like it mounts to a volume. Should we add a secret in addition to that or is there a reason we should create a new one.

Reply:
So the user has to download the key and then add it to their GITHUB repository as a deploy key? Is that documented somewhere.

Reply:
So I am looking here: 
https://github.com/weaveworks/flux/blob/master/site/daemon.md

and "flux-git-deploy" seems to be used to store the private SSH key as a secret for GitHub. 

This documentation (https://www.weave.works/docs/cloud/latest/tasks/deploy/manual-configuration/) suggests that one use "fluxctl identity" to generate a public key while the private key is stored in "flux-git-deploy". You then go to Github and add the Key in the settings.